onnx/defs/schema.h

```diff
@@ -33,6 +33,18 @@ typedef std::unordered_set<DataType> DataTypeSet;
 typedef std::unordered_map<std::string, std::pair<DataTypeSet, std::string>>
     TypeConstraintMap;
 
+typedef TensorShapeProto_Dimension InferenceDimension;
+
+struct InferenceContext {
+  virtual const AttributeProto* getAttribute(const std::string& name) const = 0;
+  virtual size_t getNumInputTypes() const = 0;
+  virtual const TypeProto_Tensor* getInputType(size_t index) const = 0;
+  virtual size_t getNumOutputTypes() const = 0;
+  virtual TypeProto_Tensor* getOutputType(size_t index) = 0;
+};
```

Comment:
Thanks for initiating the shape inference work.

One suggestion: I think it will be useful to extend “InferenceContext” by adding a method that allows the shape-inference function to generate a new symbol (name) representing a new unknown dimension. e.g.:
	virtual string GenerateNewDimParam() = 0;

Explanation: TensorShapeProto::Dimension allows the use of “string dim_param” to denote a dimension of unknown size. If some op OP1 returns a one-dimensional tensor whose size will be known only at runtime, it shape-inference function can return (“N”) as the output shape. However, different invocations of OP1 can return tensors of different sizes. If we have two occurrences of OP1 in a graph, we would like shape-inference to return different dimensions for the two cases: say (“N1”) and (“N2”). Only the caller has the necessary context to ensure uniqueness of these names.

Reply:
Yes, that could be useful. An alternative is being discussed here https://github.com/onnx/onnx/pull/681#discussion_r178886494 - reserve the empty string as a special value that is treated as a unique name

Reply:
That's true. I'll probably add it at some point then - if you have a more urgent need feel free to also to send a PR with GenerateNewDimParam.